### ISSUE TITLE
Server side vote system defence

### ISSUE THREAD

**Issue by ForNeVeR** [codingteam/loglist#103]
There should be a server-side block when a user attempts to spam our voting system.

**Comment by ForNeVeR** [comment]
I've discussed the question with my colleagues, and we've come up with the following solution: we could ask our user to solve the CAPTCHA first time he votes for something. After that we can store the received (secret) key in some client storage like cookies or JS memory or localstorage or whatever.

So, ordinary users will be asked to solve the CAPTCHA to vote, but they will be asked to do it only once (per session or per whatever period we'll choose as a validity window for our key; probably we won't want to invalidate it too often, maybe even never).

@rexim, could you approve that idea in principle? If so, I'll come up with additional technical details in the nearest days.

**Comment by ForNeVeR** [comment]
Note that the resulting option will serve as both the client-side and the server-side solution (at least for browser client - we'll need to invent something better for other client types after we'll have them), so maybe we'll merge #102 and #103 together again, as they were back the days of #37 :)

**Comment by rexim** [comment]
It's an interesting idea, but I need to think about it. I'll give my answer once I'm home. Thanks!

**Comment by ForNeVeR** [comment]
No problems, I'm ready to discuss it through more operative channels if you wish.

**Comment by hagane** [comment]
@ForNeVeR There's a problem with this: I can solve your captcha, obtain a key and then target a LOIC with that key onto your up- or downvote link.

I think a rudimentary sort-of-proof-of-work system would be much better. What i propose is: when user wants to vote, we give them a random hex string and a piece of javascript code performing some calculation with that string as an input. The latter can be embedded in the page itself. The client performs this calculation, and sends the result along with the vote request. Server checks this result and approves or denies that vote. Work unit for a quote should be constant fot the session and the server should disallow subsequent requests with that work unit, of course.

**Comment by ForNeVeR** [comment]
> I can solve your captcha, obtain a key and then target a LOIC with that key onto your up- or downvote link.

Yes you can, but you cannot do more than any ordinary user (except from eating our traffic, but that's the whole different story). We **should** remember the vote history for every captcha-received key and don't allow it to vote more than once per day. I implied this when was suggesting the idea.

**Comment by hagane** [comment]
> don't allow it to vote more than once per day.

So I can't upvote two ore more quotes in one day?

**Comment by ForNeVeR** [comment]
> The client performs this calculation, and sends the result along with the vote request. Server checks this result and approves or denies that vote. Work unit for a quote should be constant fot the session and the server should disallow subsequent requests with that work unit, of course.

First of all, it will not work with disabled JS, but we wish it work. Also, I cannot see how it solves the problem. The user can write the client that will execute our JS (and it even will be desirable for him because there're many kiddies who can into JS these days). I don't think we could and should invest into the system tough enough to be defended from these kiddies.

(Here I imply that it's much harder for hacker to write CAPTCHA-solving system than execute arbitary JS.)

**Comment by ForNeVeR** [comment]
> So I can't upvote two ore more quotes in one day?

So you can't upvote one quote more that once per day. But you may upvote (downvote) **every** quote in one day. That's completely possible and we won't apply any measures to users (_incl. client-writers_) who use this ability for their good.

**Comment by hagane** [comment]
Let's discuss it later today, in a MUC or Gitter or something.

**Comment by ForNeVeR** [comment]
Agreed.
